File: examples/python/agent/local_tools.py
"""Demo agent: LangGraph agent with local tool execution (no runner needed)."""

import ast
import asyncio
import logging
import operator
import os
from typing import Any

from langchain_openai import ChatOpenAI
from langchain.agents import create_agent

from rebuno import AsyncAgentContext, AsyncBaseAgent

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(name)s %(levelname)s %(message)s"
)
logger = logging.getLogger("demo-agent-local-tools")

SYSTEM_PROMPT = (
    "You are a research assistant. You have access to tools for searching the web, "
    "fetching documents, and doing math. Use them to answer the user's question. "
    "When you have enough information, provide a clear final answer."
)


class ResearchAgent(AsyncBaseAgent):
    """LangGraph agent where tools run in the agent process.

    Same LLM reasoning as the remote-tools demo, but tools execute in-process
    instead of being dispatched to a separate runner. The kernel still
    enforces policy and records every tool call as events.
    """

    def __init__(self, **kwargs: Any):
        super().__init__(**kwargs)
        self._model = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")

    async def process(self, ctx: AsyncAgentContext) -> dict:
        query = ""
        if isinstance(ctx.input, dict):
            query = ctx.input.get("query", "")
        elif isinstance(ctx.input, str):
            query = ctx.input

        if not query:
            return {"error": "No query provided"}

        logger.info("Processing query: %s", query)

        # ctx.get_tools() returns wrapped callables with the original
        # signatures and docstrings — any framework can consume them directly.
        tools = ctx.get_tools()

        llm = ChatOpenAI(model=self._model, temperature=0)
        agent = create_agent(
            model=llm,
            tools=tools,
            system_prompt=SYSTEM_PROMPT,
        )

        result = await agent.ainvoke(
            {"messages": [{"role": "user", "content": query}]}
        )

        final_message = result["messages"][-1].content
        logger.info("Agent finished: %s", final_message)

        return {"query": query, "answer": final_message}


_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}


def _safe_eval(expr: str) -> float:
    tree = ast.parse(expr, mode="eval")
    return _eval_node(tree.body)


def _eval_node(node: ast.AST) -> float:
    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
        return node.value
    if isinstance(node, ast.BinOp) and type(node.op) in _OPS:
        return _OPS[type(node.op)](_eval_node(node.left), _eval_node(node.right))
    if isinstance(node, ast.UnaryOp) and type(node.op) in _OPS:
        return _OPS[type(node.op)](_eval_node(node.operand))
    raise ValueError(f"Unsupported expression: {type(node).__name__}")


agent = ResearchAgent(
    agent_id="researcher-local",
    kernel_url=os.environ.get("REBUNO_KERNEL_URL", "http://localhost:8080"),
    api_key="your-secret-token",
)

# Register tools with typed signatures. The SDK preserves the function
# name, docstring, and type annotations so that frameworks like LangGraph
# can introspect them to generate schemas for the LLM.

@agent.tool("web.search")
async def web_search(query: str) -> dict:
    """Search the web for information about a topic."""
    logger.info("Local web search: %s", query)
    await asyncio.sleep(0.2)
    return {
        "query": query,
        "results": [
            {"title": f"Result for: {query}", "snippet": f"Info about {query}."},
            {"title": f"More on: {query}", "snippet": f"Details about {query}."},
        ],
    }

@agent.tool("doc.fetch")
async def doc_fetch(url: str) -> dict:
    """Fetch and read the contents of a document at a URL."""
    logger.info("Local doc fetch: %s", url)
    await asyncio.sleep(0.1)
    return {
        "url": url,
        "title": f"Document from {url}",
        "content": f"Mock content fetched from {url}.",
        "word_count": 42,
    }

@agent.tool("calculator")
def calculator(expression: str) -> dict:
    """Evaluate a mathematical expression. Example: '2 + 3 * 4'."""
    logger.info("Local calculator: %s", expression)
    result = _safe_eval(expression)
    return {"expression": expression, "result": result}


if __name__ == "__main__":
    asyncio.run(agent.run())

File: examples/typescript/agent/local_tools.ts
// Demo agent: LangGraph agent with local tool execution (no runner needed).

import { z } from "zod";
import { ChatOpenAI } from "@langchain/openai";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { BaseAgent, AgentContext, defineTool } from "rebuno";
import { toLangchainTools } from "rebuno/tools/adapters/langchain";

const SYSTEM_PROMPT =
  "You are a research assistant. You have access to tools for searching the web, " +
  "fetching documents, and doing math. Use them to answer the user's question. " +
  "When you have enough information, provide a clear final answer.";

// LangGraph agent where tools run in the agent process.
//
// Same LLM reasoning as the remote-tools demo, but tools execute in-process
// instead of being dispatched to a separate runner. The kernel still
// enforces policy and records every tool call as events.
class ResearchAgent extends BaseAgent {
  private model = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

  async process(ctx: AgentContext) {
    const input = ctx.input as Record<string, unknown> | string;
    const query =
      typeof input === "object" && input !== null
        ? (input.query as string) ?? ""
        : String(input);

    if (!query) return { error: "No query provided" };

    console.log(`Processing query: ${query}`);

    // ctx.getTools() returns wrapped tools — adapters convert them to
    // framework-specific formats (here LangChain StructuredTool).
    const tools = await toLangchainTools(ctx.getTools()) as any;

    const llm = new ChatOpenAI({ model: this.model, temperature: 0 });
    const agent = createReactAgent({
      llm,
      tools,
      stateModifier: SYSTEM_PROMPT,
    });

    const result = await agent.invoke({
      messages: [{ role: "user", content: query }],
    });

    const finalMessage = result.messages[result.messages.length - 1].content;
    const answer =
      typeof finalMessage === "string"
        ? finalMessage
        : JSON.stringify(finalMessage);
    console.log(`Agent finished: ${answer.slice(0, 200)}`);

    return { query, answer };
  }
}

function safeEval(expr: string): number {
  const tokens = expr.match(/(\d+\.?\d*|[+\-*/()])/g);
  if (!tokens) throw new Error("Empty expression");
  let pos = 0;

  function parseExpr(): number {
    let result = parseTerm();
    while (pos < tokens!.length && (tokens![pos] === "+" || tokens![pos] === "-")) {
      const op = tokens![pos++];
      const right = parseTerm();
      result = op === "+" ? result + right : result - right;
    }
    return result;
  }

  function parseTerm(): number {
    let result = parseFactor();
    while (pos < tokens!.length && (tokens![pos] === "*" || tokens![pos] === "/")) {
      const op = tokens![pos++];
      const right = parseFactor();
      result = op === "*" ? result * right : result / right;
    }
    return result;
  }

  function parseFactor(): number {
    if (tokens![pos] === "(") {
      pos++;
      const result = parseExpr();
      pos++; // skip ')'
      return result;
    }
    if (tokens![pos] === "-") {
      pos++;
      return -parseFactor();
    }
    return parseFloat(tokens![pos++]);
  }

  return parseExpr();
}

const agent = new ResearchAgent({
  agentId: "researcher-local",
  kernelUrl: process.env.REBUNO_KERNEL_URL ?? "http://localhost:8080",
});

// Register tools with defineTool(). The SDK uses the id, description, and
// inputSchema so that frameworks like LangGraph can generate schemas for the LLM.

agent.addTool(
  defineTool({
    id: "web.search",
    description: "Search the web for information about a topic.",
    inputSchema: z.object({ query: z.string() }),
    execute: async ({ query }) => {
      console.log(`Local web search: ${query}`);
      await new Promise((r) => setTimeout(r, 200));
      return {
        query,
        results: [
          { title: `Result for: ${query}`, snippet: `Info about ${query}.` },
          { title: `More on: ${query}`, snippet: `Details about ${query}.` },
        ],
      };
    },
  }),
);

agent.addTool(
  defineTool({
    id: "doc.fetch",
    description: "Fetch and read the contents of a document at a URL.",
    inputSchema: z.object({ url: z.string() }),
    execute: async ({ url }) => {
      console.log(`Local doc fetch: ${url}`);
      await new Promise((r) => setTimeout(r, 100));
      return {
        url,
        title: `Document from ${url}`,
        content: `Mock content fetched from ${url}.`,
        wordCount: 42,
      };
    },
  }),
);

agent.addTool(
  defineTool({
    id: "calculator",
    description: "Evaluate a mathematical expression. Example: '2 + 3 * 4'.",
    inputSchema: z.object({ expression: z.string() }),
    execute: async ({ expression }) => {
      console.log(`Local calculator: ${expression}`);
      const result = safeEval(expression);
      return { expression, result };
    },
  }),
);

agent.run();
